syntax = "proto3";
package streetlevel;

import "Shared.proto";

message PlaceResponse {
  int32 status = 1;
  int32 request_type = 2;
  PlaceGlobalResult global_result = 3;
  //Place legacy_place_result = 4;
  repeated string display_language = 5;
  repeated string display_region = 6;
  repeated string spoken_language = 7;
  string debug_api_key = 8;
  DatasetABStatus dataset_ab_status = 9;
  MapsResult maps_result = 10;
  //DotPlace dot_place = 11;
  uint64 debug_latency_ms = 1000;
  //ClientMetrics client_metrics = 1051;

  message PlaceGlobalResult {
    // TODO everything else
    ReverseGeocodingResult reverse_geocoding_result = 4;
    message ReverseGeocodingResult {
      int32 unknown = 100;
    }
  }

  message DatasetABStatus {
    uint32 dataset_id = 1;
  }

  message MapsResult {
    MapsResultType result_type = 1;
    Place place = 2;
    //CollectionResult collection = 3;
    //PublisherResult publisher = 4;
    //PlaceQuestionnaireResult place_questionnaire_result = 100;
    //BatchReverseGeocodeResult batch_reverse_geocode = 101;
  }

  message Place {
    uint64 muid = 1;
    Status status = 2;
    uint64 preferred_muid = 3;
    repeated Component component = 4;
    int32 result_provider_id = 5;
    uint64 update_version = 6;
    MapsIdentifier maps_id = 7;
    uint64 created_time = 8;
    //PlacecardLayoutData place_layout_data = 9;
    //TODO the rest of this
  }

  message Component {
    int32 type = 1;
    int32 status = 2;
    int32 ttl = 4;
    int32 start_index = 5;
    bool values_available = 6;
    //Attribution attribution = 7;
    repeated ComponentValue value = 8;
    repeated string version_domain = 9;
    int32 version = 10;
    //Source source = 11;
    int32 cache_control = 12;
    uint64 component_id = 13;
    //repeated Attribution commingled_attribution = 14;
  }

  message ComponentValue {
    // TODO everything else
    AddressObject address_object = 31;
  }

  message AddressObject {
    // There's some weird stuff going on with this field.
    // I think it's actually defined as `bytes` in the original .proto
    // and it gets parsed in some other place. In any case, I haven't been
    // able to recover the original identifiers for this message, so I'm just
    // going to call `AddressObject_` for now, and all identifiers of its
    // members are my own
    AddressObject_ address_object = 1;
  }

  message AddressObject_ {
    int32 unknown1 = 1;
    string unknown2 = 2;
    Unknown3 unknown3 = 3;
    Unknown4 unknown4 = 4;
    Address address = 101;
  }

  message Unknown3 {
    string unknown5 = 1;
    string unknown6 = 2;
    string unknown7 = 3;
  }

  message Unknown4 {
    string unknown8 = 1;
    uint32 unknown9 = 2;
    string unknown10 = 3;
    string unknown11 = 4;
  }

  message Address {
    string place_1 = 1;
    Address1 address = 2;
    string place_2 = 3;
    string formatted_place = 4;
    FormattedAddressComponents formatted_address_components = 5;
    string unknown14 = 7;
    Unknown15 unknown15 = 100;
  }

  message Address1 {
    repeated string address_line = 11;
    AddressComponents address_components = 15;
  }

  message AddressComponents {
    string country = 1;
    string country_code = 2; //  ISO 3166 country code
    string state = 3;
    string region_code = 4; // ISO 3166 region code
    string county = 5;
    string city = 6;
    string postal_code = 7;
    string neighborhood = 8;
    string street = 10;
    string house_number = 11;
    string street_and_house_number = 12;
    repeated string place = 14;
    repeated string enclosing_area = 17;
  }

  message FormattedAddressComponents {
    string country_1 = 1;
    string country_2 = 2;
    string state_1 = 3;
    string state_2 = 4;
    string county = 5;
    string city = 6;
    string neighborhood_1 = 7;
    string neighborhood_2 = 8;
    string street = 10;
    string house_number = 11;
    string street_and_house_number = 12;
    repeated string place = 14;
    repeated string enclosing_area = 17;
  }

  message Unknown15 {
    string city_1 = 1;
    string place = 2;
    string street_and_city = 5;
    string city_2 = 6;
    string ward = 10;

  }

  message MapsIdentifier {
    ShardedId sharded_id = 1;
  }

  message ShardedId {
    uint64 muid = 1;
    LatLng center = 2;
    int32 result_provider_id = 3;
    int32 basemap_id = 4;
    int32 unknown0 = 5;
  }
}

enum Status {
  STATUS_SUCCESS = 0;
  STATUS_FAILED = 1;
  STATUS_INCOMPLETE = 2;
  INVALID_REQUEST = 5;
  FAILED_NO_RESULT = 20;
  NEEDS_REFINEMENT = 30;
  FAILED_NOT_AUTHORIZED = 40;
  STATUS_DEDUPED = 50;
  VERSION_MISMATCH = 60;
}
